src/java.base/share/classes/jdk/internal/vm/vector/VectorSupport.java

```diff
@@ -453,6 +458,20 @@ VOUT convert(int oprId,
         return v;
     }
 
+    /* ============================================================================ */
+    public interface VectorMaskOp<M> {
+        int apply(M m);
+    }
+
+    @IntrinsicCandidate
+    public static
+    <M>
+    int maskOp(int oper, Class<?> maskClass, Class<?> elemClass, int length, M m,
```

Comment:
As discussed above mixing it with reduction coerced will require changes in original entry point (type parameter have Vector as the lower bound) , also we may need to bypass some irrelevant portions in inline_vector_reduction() , for the time being to keep the things clean added a different entry point for all masked operations.

Reply:
Ok, fair enough. We can revisit that later and merge them if needed.
Some suggestions to consider to align it with `reductionCoerced`:
* reflect in the name that it's effectively a reduction, but on masks (`maskReductionCoerced`?);
* return type can be generalized to `long`;
* bound on M: `<M extends VectorMask>`;
* no need to introduce a special interface, `Function<T,R>` just works: `VectorMaskOp<M>` -> `Function<M, Long>`;

Reply:
> Ok, fair enough. We can revisit that later and merge them if needed.
> Some suggestions to consider to align it with `reductionCoerced`:
> 
> * reflect in the name that it's effectively a reduction, but on masks (`maskReductionCoerced`?);
> * return type can be generalized to `long`;

Hi @iwanowww,  Can you kindly elaborate why should the return type be long here ? 
We will need to again downcast it to integer since these APIs return an integer value.

> * bound on M: `<M extends VectorMask>`;
> * no need to introduce a special interface, `Function<T,R>` just works: `VectorMaskOp<M>` -> `Function<M, Long>`;